## 🎯 Objetivos del estudio

El propósito de este estudio es analizar la base de datos de una startup del sector editorial que ofrece un servicio digital para lectores. A raíz del incremento en el consumo de libros durante la pandemia, muchas empresas emergentes han comenzado a competir por atraer a este nuevo público lector.

### Objetivo general:
Desarrollar una **propuesta de valor basada en datos** que permita identificar oportunidades de mejora o expansión para el producto digital ofrecido por la empresa.

### Objetivos específicos:

- 📚 Identificar **los libros, autores y editoriales más populares**, tanto en términos de volumen como de calidad percibida por los usuarios.
- ⭐ Analizar el comportamiento de los usuarios mediante calificaciones y reseñas, identificando **lectores activos** y sus patrones.
- 🔍 Detectar **contenidos relevantes** (libros con alta participación, buena valoración, o autores con impacto) que puedan ser destacados en campañas o nuevas funcionalidades del producto.
- 🚫 Excluir publicaciones marginales (como folletos) mediante criterios de calidad como número mínimo de páginas o calificaciones.

Este análisis se realizará mediante consultas SQL, explorando la base de datos relacional proporcionada, que incluye información sobre libros, autores, editoriales, calificaciones y reseñas de usuarios.


In [2]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-final-project-db'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

In [3]:
query = 'SELECT * FROM books LIMIT 5;'
books_df = pd.read_sql(query, con=engine)

books_df

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


In [4]:
table_names = ['books', 'authors', 'publishers', 'ratings', 'reviews']

for table in table_names:
    print(f'\n🔹 Primeras filas de la tabla {table}:\n')
    display(pd.read_sql(f'SELECT * FROM {table} LIMIT 5;', con=engine))


🔹 Primeras filas de la tabla books:



,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268



🔹 Primeras filas de la tabla authors:



,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd



🔹 Primeras filas de la tabla publishers:



,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company



🔹 Primeras filas de la tabla ratings:



,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2



🔹 Primeras filas de la tabla reviews:



,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


## 🔍 Estudio preliminar de las tablas

Para familiarizarnos con la estructura y el contenido de la base de datos, se consultaron las primeras 5 filas de cada una de las tablas disponibles. A continuación se describen los resultados:

### 📚 Tabla `books`
Contiene información básica de los libros como título, número de páginas, fecha de publicación, autor y editorial asociada.

| book_id | author_id | title                     | num_pages | publication_date | publisher_id |
|---------|-----------|---------------------------|-----------|------------------|---------------|
| 1       | 546       | Salem's lot               | 594       | 2005-11-01       | 93            |
| 2       | 214       | 1000 Place to see befo... | 992       | 2003-05-22       | 336           |
| ...     | ...       | ...                       | ...       | ...              | ...           |

---

### ✍️ Tabla `authors`
Contiene una lista de autores, vinculados a sus libros mediante `author_id`.

| author_id | author                        |
|-----------|-------------------------------|
| 1         | A.S Byatt                     |
| 2         | Aesop/laura harris/Laura Gibbs|
| ...       | ...                           |

---

### 🏢 Tabla `publishers`
Lista de editoriales disponibles.

| publisher_id | publisher                |
|--------------|--------------------------|
| 1            | Ace                      |
| 2            | Ace books                |
| ...          | ...                      |

---

### ⭐ Tabla `ratings`
Contiene las calificaciones de los usuarios a los libros.

| rating_id | book_id | username     | rating |
|-----------|---------|--------------|--------|
| 1         | 1       | ryanfranco   | 4      |
| 2         | 1       | reader_kyle  | 2      |
| ...       | ...     | ...          | ...    |

---

### 📝 Tabla `reviews`
Contiene las reseñas textuales de los usuarios.

| review_id | book_id | username      | text                                             |
|-----------|---------|---------------|--------------------------------------------------|
| 1         | 1       | brandtandrea  | "Mention society tell send professor analysis....|
| 2         | 1       | ryanfranco    | "Foot glass pretty audience hit themselves. Amo..|
| ...       | ...     | ...           | ...                                              |

---

Estas consultas permiten confirmar que las tablas están correctamente relacionadas mediante claves foráneas (`author_id`, `publisher_id`, `book_id`) y que los datos están estructurados de forma consistente para realizar el análisis exploratorio y analítico.


In [5]:
query = '''
SELECT COUNT(*) AS books_after_2000
FROM books
WHERE publication_date > '2000-01-01';
'''

books_after_2000 = pd.read_sql(query, con=engine)

books_after_2000


,books_after_2000
0,819


### ✅ Conclusión: Libros publicados después del año 2000

Se identificaron **819 libros** en la base de datos cuya fecha de publicación es posterior al **1 de enero del año 2000**.

Este hallazgo es relevante porque:

- Representa una gran cantidad de contenido **moderno y potencialmente más relevante** para los lectores actuales.
- Nos permite enfocar análisis posteriores (como calificaciones, reseñas y autores destacados) en libros **más recientes**, lo cual puede mejorar la propuesta de valor del producto digital.

📌 *Este subconjunto puede ser utilizado para análisis filtrados donde se desee trabajar solo con publicaciones del siglo XXI.*

In [6]:
query = '''
SELECT
    b.book_id,
    b.title,
    COUNT(DISTINCT r.review_id) AS review_count,
    ROUND(AVG(rt.rating), 2) AS avg_rating
FROM books b
LEFT JOIN reviews r ON b.book_id = r.book_id
LEFT JOIN ratings rt ON b.book_id = rt.book_id
GROUP BY b.book_id, b.title
ORDER BY review_count DESC;
'''

book_reviews_ratings = pd.read_sql(query, con=engine)

book_reviews_ratings.head(10)  

,book_id,title,review_count,avg_rating
0,948,Twilight (Twilight #1),7,3.66
1,963,Water for Elephants,6,3.98
2,734,The Glass Castle,6,4.21
3,302,Harry Potter and the Prisoner of Azkaban (Harr...,6,4.41
4,695,The Curious Incident of the Dog in the Night-Time,6,4.08
5,696,The Da Vinci Code (Robert Langdon #2),6,3.83
6,627,The Alchemist,6,3.79
7,750,The Hobbit or There and Back Again,6,4.13
8,656,The Book Thief,6,4.26
9,779,The Lightning Thief (Percy Jackson and the Oly...,6,4.08


### ✅ Conclusión: Número de reseñas y calificación promedio por libro

Se realizó un análisis combinado de las tablas `reviews` y `ratings` para identificar qué libros tienen mayor participación por parte de los usuarios, tanto en forma de reseñas como de calificaciones.

#### 📌 Hallazgos destacados:
- 📖 **"Twilight (Twilight #1)"** es el libro con mayor número de reseñas (7), aunque su calificación promedio es relativamente baja (**3.66**).
- 📖 **"Harry Potter and the Prisoner of Azkaban"** destaca por tener una excelente calificación promedio (**4.41**) con 6 reseñas, mostrando alto nivel de satisfacción.
- Otros libros con **buen equilibrio entre cantidad de reseñas y buena valoración** incluyen:
  - *The Book Thief* (4.26)
  - *The Glass Castle* (4.21)
  - *The Hobbit* (4.13)

Estos resultados son útiles para:
- Identificar libros que **generan conversación** (muchas reseñas).
- Detectar títulos con **alto potencial de recomendación** (alta calificación).
- Priorizar obras con buena recepción para **recomendaciones automáticas**, promociones o curación de contenido en el producto digital.



In [11]:
query = '''
SELECT
    p.publisher,
    COUNT(b.book_id) AS book_count
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY book_count DESC
LIMIT 1;
'''

top_publisher = pd.read_sql(query, con=engine)

top_publisher


,publisher,book_count
0,Penguin Books,42


### ✅ Conclusión: Editorial con más libros de más de 50 páginas

Se realizó un análisis para identificar la editorial que ha publicado la mayor cantidad de libros **con más de 50 páginas**, con el objetivo de filtrar folletos o publicaciones de poca extensión.

#### 📌 Resultado:
- 🏢 **Penguin Books** es la editorial líder, con **42 libros** publicados que superan las 50 páginas.

Este hallazgo es relevante porque:

- Permite enfocarse en editoriales que **apuestan por contenido más extenso y posiblemente de mayor profundidad**.
- Penguin Books puede representar un **socio clave** en futuras estrategias de contenido para la plataforma.
- Filtrar por número de páginas es una buena práctica para mejorar la **calidad de las recomendaciones y análisis**, excluyendo entradas poco significativas.


In [14]:
query = '''
WITH rated_books AS (
    SELECT
        b.book_id,
        b.title,
        b.author_id,
        COUNT(r.rating_id) AS num_ratings,
        AVG(r.rating) AS avg_rating
    FROM books b
    JOIN ratings r ON b.book_id = r.book_id
    GROUP BY b.book_id
    HAVING COUNT(r.rating_id) >= 50
)

SELECT
    a.author,
    rb.title,
    ROUND(rb.avg_rating, 2) AS avg_rating,
    rb.num_ratings
FROM rated_books rb
JOIN authors a ON rb.author_id = a.author_id
ORDER BY rb.avg_rating DESC
LIMIT 1;
'''

top_rated_author = pd.read_sql(query, con=engine)

top_rated_author


,author,title,avg_rating,num_ratings
0,J.K. Rowling/Mary GrandPré,Harry Potter and the Prisoner of Azkaban (Harr...,4.41,82


### ✅ Conclusión: Autor con la calificación promedio más alta (mín. 50 calificaciones)

Se realizó un análisis para identificar al autor o autora cuyo libro tiene la **calificación promedio más alta**, considerando únicamente libros que hayan recibido **al menos 50 calificaciones**. Esta restricción asegura que el promedio sea estadísticamente confiable.

#### 📌 Resultado:
- ✍️ **Autor(a):** J.K. Rowling / Mary GrandPré  
- 📖 **Libro:** *Harry Potter and the Prisoner of Azkaban*  
- ⭐ **Promedio de calificación:** 4.41  
- 🧮 **Número de calificaciones:** 82

Este hallazgo destaca a **J.K. Rowling** como una de las autoras más reconocidas y valoradas por los lectores de la plataforma, especialmente dentro de la literatura juvenil y fantástica.


In [13]:
query = '''
WITH active_users AS (
    SELECT
        username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 50
),
user_reviews AS (
    SELECT
        r.username,
        COUNT(rv.review_id) AS num_reviews
    FROM active_users r
    LEFT JOIN reviews rv ON r.username = rv.username
    GROUP BY r.username
)

SELECT
    ROUND(AVG(num_reviews), 2) AS avg_reviews_per_active_user
FROM user_reviews;
'''

avg_reviews = pd.read_sql(query, con=engine)

avg_reviews


,avg_reviews_per_active_user
0,24.33


### ✅ Conclusión: Promedio de reseñas escritas por usuarios que calificaron más de 50 libros

Se analizó la relación entre los usuarios que han calificado una gran cantidad de libros (más de 50) y el número de reseñas textuales que escriben.

#### 📌 Resultado:
- Los usuarios que calificaron más de 50 libros escribieron, en promedio, **24.33 reseñas de texto**.

Este dato es útil porque:

- Muestra que los **usuarios muy activos no solo califican, sino que también contribuyen con contenido textual**, lo cual es valioso para enriquecer la plataforma.
- Las reseñas de texto ofrecen contexto y opiniones más ricas que una simple calificación, ayudando a otros usuarios en su decisión de lectura.
- Estos usuarios comprometidos podrían ser candidatos ideales para **programas de fidelización**, acceso anticipado a libros o funciones exclusivas dentro del producto.

## 🧾 Conclusión Final del Proyecto

A través del análisis de la base de datos proporcionada por una startup del sector editorial, se exploraron múltiples aspectos relacionados con los libros, autores, editoriales y el comportamiento de los usuarios en términos de calificaciones y reseñas.

### 🔍 Hallazgos clave:

- Se identificaron **819 libros publicados después del año 2000**, lo que indica un catálogo moderno y relevante.
- El libro con mayor número de reseñas fue *"Twilight (Twilight #1)"*, mientras que *"Harry Potter and the Prisoner of Azkaban"* destacó como el libro mejor calificado con más de 50 valoraciones (**4.41** de promedio).
- La editorial **Penguin Books** lidera en volumen de publicaciones con más de 50 páginas (**42 libros**), lo que la convierte en un socio estratégico potencial.
- Usuarios altamente activos (que calificaron más de 50 libros) escriben en promedio **24.33 reseñas**, mostrando un fuerte compromiso con la plataforma.

### 📌 Recomendaciones basadas en el análisis:

- **Fomentar la visibilidad** de libros y autores mejor calificados para impulsar el engagement.
- **Aprovechar editoriales líderes** como Penguin Books para enriquecer la oferta de contenido.
- **Incentivar a los usuarios activos**, ya que generan contenido valioso y confiable para la comunidad.
- **Filtrar publicaciones irrelevantes** (como folletos cortos) mediante criterios como número de páginas para mejorar la calidad del análisis.

---

Este análisis proporciona una base sólida para definir una propuesta de valor orientada a:
- Mejorar las recomendaciones personalizadas,
- Promover autores y libros con alto impacto,
- Reforzar la fidelización de usuarios activos,
- Y enriquecer la estrategia de contenido del producto digital.
